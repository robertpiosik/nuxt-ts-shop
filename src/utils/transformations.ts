export const priceFormatter = (price: number) => {
  let newPrice = price.toString()

  const newPriceSplitted = newPrice.split('.')

  const value = newPriceSplitted[0]
  let penny: string

  if (newPriceSplitted[1]) {
    if (newPriceSplitted[1].length >= 2) {
      penny = `,${newPriceSplitted[1].slice(0, 2)}`
    } else {
      penny = `,${newPriceSplitted[1]}0`
    }
  } else {
    penny = `,00`
  }

  return {
    value,
    penny
  }
}
