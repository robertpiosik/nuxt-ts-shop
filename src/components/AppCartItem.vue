<template>
  <div :class="$style.container">
    <div :class="$style.thumbnailAndTitle">
      <div :class="$style.thumbnail">
        <img :src="thumbnail" :alt="title" ref="thumbnailImage" />
      </div>
      <div :class="$style.titleAndRemove">
        <div :class="$style.title">{{ title }}</div>
        <div>
          <button :class="$style.remove" @click="removeItem()" remove="remove">
            &#10005; Remove
          </button>
        </div>
      </div>
    </div>
    <div :class="$style.quantity">
      <div :class="$style.quantityAmount">{{ quantity }}</div>
      <button
        :class="[$style.quantityButton, $style.quantityButtonDecrease]"
        @click="decreaseQuantity()"
        decreaseQuantity="decreaseQuantity"
      >
        -</button
      ><button
        :class="[$style.quantityButton, $style.quantityButtonIncrease]"
        @click="increaseQuantity()"
        increaseQuantity="increaseQuantity"
      >
        +
      </button>
    </div>
    <div :class="$style.priceAndTotal">
      <template v-if="price !== price * quantity">
        <div>{{ `${priceFormatted.value + priceFormatted.penny} zł` }}</div>
        <div>
          <strong>{{
            `${totalFormatted.value + totalFormatted.penny} zł`
          }}</strong>
        </div>
      </template>
      <template v-else>
        <div>
          <strong>{{
            `${priceFormatted.value + priceFormatted.penny} zł`
          }}</strong>
        </div>
      </template>
    </div>
  </div>
</template>

<script lang="ts">
import { Vue, Component, Prop, Emit, Ref } from 'vue-property-decorator';

import { priceFormatter } from './../utils/transformations';

@Component
export default class CartItem extends Vue {
  @Prop({ type: String, required: true }) readonly title!: string;
  @Prop({ type: Number, required: true }) readonly price!: number;
  @Prop({ type: String, required: true }) readonly thumbnail!: string;
  @Prop({ type: Number, required: true }) readonly quantity!: number;

  get priceFormatted() {
    return priceFormatter(this.price);
  }

  get totalFormatted() {
    return priceFormatter(this.price * this.quantity);
  }

  @Emit()
  increaseQuantity() {}

  @Emit()
  decreaseQuantity() {}

  @Emit()
  removeItem() {}

  @Ref('thumbnailImage') readonly thumbnailImage!: HTMLImageElement;

  mounted() {
    if (typeof (<any>window).objectFitPolyfill === 'function') {
      (<any>window).objectFitPolyfill(this.thumbnailImage);
    }
  }
}
</script>

<style lang="scss" module>
.container {
  display: grid;
  margin: 1.5rem 0;
  max-width: 65rem;
  padding-bottom: 1.5rem;
  border-bottom: 0.1rem solid rgba(black, 0.1);
  grid-template:
    'thumbnailAndTitle thumbnailAndTitle'
    'quantity priceAndTotal' / 8rem 1fr;
  gap: 1.5rem;
  @include atSmall {
    grid-template: 'thumbnailAndTitle quantity priceAndTotal' / 3fr 1fr 1fr;
  }
}
.thumbnailAndTitle {
  grid-area: thumbnailAndTitle;

  display: grid;
  grid-template-columns: 8rem auto;
  grid-template-rows: 8rem;
  gap: 1.5rem;
  @include atSmall {
    grid-template-columns: 10rem auto;
    grid-template-rows: 10rem;
  }
}
.thumbnail {
  position: relative;
  overflow: hidden;
  border-radius: 0.5rem;
  & img {
    width: 100%;
    height: 100%;
    position: absolute;
    object-fit: cover;
  }
}
.titleAndRemove {
  display: flex;
  flex-direction: column;
  justify-content: center;
}
.title {
  display: flex;
  align-items: center;
  font-weight: bold;
  color: $primary;
}
.remove {
  font-size: 1.2rem;
  font-weight: bold;
  color: rgba(black, 0.3);
  cursor: pointer;
  padding: 0.5rem 1rem 0.5rem 0;
  &:hover {
    color: red;
  }
}
.quantity {
  grid-area: quantity;
  justify-self: end;

  display: grid;
  gap: 0.5rem;
  grid-template:
    'quantityAmount quantityButtonDecrease' 1fr
    'quantityAmount quantityButtonIncrease' 1fr / auto auto;
}
.quantityAmount {
  grid-area: quantityAmount;

  align-self: center;
  font-weight: bold;
  margin-right: 1rem;
}
.quantityButtonDecrease {
  grid-area: quantityButtonDecrease;
}
.quantityButtonIncrease {
  grid-area: quantityButtonIncrease;
}
.quantityButton {
  width: 3rem;
  height: 3rem;
  display: flex;
  align-items: center;
  justify-content: center;
  border: 0.1rem solid rgba(black, 0.1);
  border-radius: 0.5rem;
  font-weight: bold;
  &:hover {
    background: rgba(black, 0.1);
  }
  &:first-of-type {
    align-self: end;
  }
}
.priceAndTotal {
  grid-area: priceAndTotal;

  display: flex;
  flex-direction: column;
  justify-content: center;
}
</style>
