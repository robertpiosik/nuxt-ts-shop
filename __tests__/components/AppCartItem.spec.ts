import { mount } from '@vue/test-utils'

import AppCartItem from './../../src/components/AppCartItem.vue'

import { priceFormatter } from './../../src/utils/transformations'

const title = 'Water'
const price = 1.2
const thumbnail = 'someUrl'
const quantity = 2

const wrapper = mount(AppCartItem, {
  propsData: {
    title,
    price,
    thumbnail,
    quantity
  }
})

describe('Component: CartItem', () => {
  describe('GENERAL', () => {
    it('should match a snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })
    it('should be a Vue instance', () => {
      expect(wrapper.isVueInstance).toBeTruthy
    })
  })
  describe('PROPS', () => {
    describe('title', () => {
      it('should receive', () => {
        expect(wrapper.props('title')).toBe(title)
      })
      it('should render', () => {
        expect(wrapper.text()).toContain(title)
      })
    })
    describe('price', () => {
      it('should receive', () => {
        expect(wrapper.props('price')).toBe(price)
      })
      it('should NOT render in unformatted form', () => {
        expect(wrapper.text()).not.toContain(price)
      })
      it('should render in formatted form', () => {
        const formattedPrice = priceFormatter(price)
        expect(wrapper.text()).toContain(
          formattedPrice.value + formattedPrice.penny
        )
      })
    })
    describe('thumbnail', () => {
      it('should receive', () => {
        expect(wrapper.props('thumbnail')).toBe(thumbnail)
      })
      it('should output html containing given url', () => {
        expect(wrapper.html()).toContain(thumbnail)
      })
    })
    describe('quantity', () => {
      it('should receive', () => {
        expect(wrapper.props('quantity')).toBe(quantity)
      })
      it('should render', () => {
        expect(wrapper.text()).toContain(quantity)
      })
    })
  })
  describe('EVENTS', () => {
    describe('removeItem', () => {
      it('should emit upon clicking on div[remove]', () => {
        const removeBtn = wrapper.find('div[remove]')
        removeBtn.trigger('click')
        expect(wrapper.emitted('remove-item')).toBeTruthy()
      })
    })
    describe('increaseQuantity', () => {
      it('should emit upon clicking on buttton[increaseQuantity]', () => {
        const increaseQuantityButton = wrapper.find('button[increaseQuantity]')
        increaseQuantityButton.trigger('click')
        expect(wrapper.emitted('increase-quantity')).toBeTruthy()
      })
    })
    describe('decreaseQuantity', () => {
      it('should emit upon clicking on button[decreaseQuantity]', () => {
        const decreaseQuantityButton = wrapper.find('button[decreaseQuantity]')
        decreaseQuantityButton.trigger('click')
        expect(wrapper.emitted('decrease-quantity')).toBeTruthy()
      })
    })
  })
})
