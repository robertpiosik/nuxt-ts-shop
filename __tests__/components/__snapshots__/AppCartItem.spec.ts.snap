// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: CartItem GENERAL should match a snapshot 1`] = `
<div class="container">
  <div class="thumbnailAndTitle">
    <div class="thumbnail"><img src="someUrl"></div>
    <div class="titleAndRemove">
      <div class="title">Water</div>
      <div role="button" remove="remove" class="remove">✕ Remove</div>
    </div>
  </div>
  <div class="quantity">
    <div class="quantityAmount">2</div><button decreasequantity="decreaseQuantity" class="quantityButton">-</button><button increasequantity="increaseQuantity" class="quantityButton">+</button>
  </div>
  <div class="priceAndTotal">
    <div>1,2 zł</div>
    <!---->
    <div><strong>2,4 zł</strong></div>
  </div>
</div>
`;
